Reaktoro/Core/Phases.cpp

```diff
@@ -1,350 +1,366 @@
+
+#include "Phases.hpp"
+
+// Reaktoro includes
+#include <Reaktoro/Common/Exception.hpp>
+
+namespace Reaktoro {
+
+GenericPhase::GenericPhase()
+{}
+
+GenericPhase::GenericPhase(const StringList& species)
+: names(species)
+{}
+
+GenericPhase::GenericPhase(const Speciate& elements)
+: symbols(elements.symbols)
+{}
+
+GenericPhase::GenericPhase(const Speciate& elements, const Exclude& withtags)
+: symbols(elements.symbols), excludetags(withtags.tags)
+{}
+GenericPhase::~GenericPhase()
+{}
+
+auto GenericPhase::setName(String name) -> GenericPhase&
+{
+    phasename = name;
+    return *this;
+}
+
+auto GenericPhase::setStateOfMatter(StateOfMatter option) -> GenericPhase&
+{
+    stateofmatter = option;
+    return *this;
+}
+
+auto GenericPhase::setAggregateState(AggregateState option) -> GenericPhase&
+{
+    aggregatestate = option;
+    return *this;
+}
+
+auto GenericPhase::setActivityModel(const ActivityModel& model) -> GenericPhase&
+{
+    activity_model = model;
+    return *this;
+}
+
+auto GenericPhase::setIdealActivityModel(const ActivityModel& model) -> GenericPhase&
+{
+    ideal_activity_model = model;
+    return *this;
+}
+
+auto GenericPhase::named(String name) -> GenericPhase&
+{
+    return setName(name);
+}
+
+auto GenericPhase::set(StateOfMatter option) -> GenericPhase&
+{
+    return setStateOfMatter(option);
+}
+
+auto GenericPhase::set(AggregateState option) -> GenericPhase&
+{
+    return setAggregateState(option);
+}
+
+auto GenericPhase::set(const ActivityModel& model) -> GenericPhase&
+{
+    return setActivityModel(model);
+}
+
+auto GenericPhase::name() const -> String
+{
+    return phasename;
+}
+
+auto GenericPhase::stateOfMatter() const -> StateOfMatter
+{
+    return stateofmatter;
+}
+
+auto GenericPhase::aggregateState() const -> AggregateState
+{
+    return aggregatestate;
+}
+
+auto GenericPhase::species() const -> const Strings&
+{
+    return names;
+}
+
+auto GenericPhase::elements() const -> const Strings&
+{
+    return symbols;
+}
+
+auto GenericPhase::activityModel() const -> const ActivityModel&
+{
+    return activity_model;
+}
+
+auto GenericPhase::idealActivityModel() const -> const ActivityModel&
+{
+    return ideal_activity_model;
+}
+
+auto GenericPhase::convert(const Database& db, const Strings& elements) const -> Phase
+{
+    error(aggregatestate == AggregateState::Undefined,
+        "GenericPhase::convert requires an AggregateState value to be specified.\n"
+        "Use method GenericPhase::setAggregateState to fix this.");
+
+    auto species = db.speciesWithAggregateState(aggregatestate);
+
+    species =
+        names.size() ? species.withNames(names) :
+        symbols.size() ? species.withElements(symbols) :
+            species.withElements(elements);
+
+    Phase phase;
+    phase = phase.withName(phasename);
+    phase = phase.withStateOfMatter(stateofmatter);
+
+    if(aggregatestate == AggregateState::Aqueous and excludetags.size())
```

Comment:
`GenericPhase` should be implemented in a generic way, without special logic for specific phase types.